### PR TITLE
Move options to pandoc-defaults.yaml

### PR DESCRIPTION
**Pull request by dhimmel**
Pandoc has introduced a `--defaults` option to "specify a set of default option settings". The benefits for us are a few-fold:

- reduce duplication of pandoc options between various build commands
- move to more of a configuration file model with the eventual goal of being able to build repositories with only content files and nothing else.

### CODE REVIEW

**Comment by dhimmel** [comment]
Here's a local WeasyPrint build of https://github.com/manubot/rootstock/pull/290/commits/6486c3be806f34cb5944a32bdaf28815cee8f462: [manuscript.pdf](https://github.com/manubot/rootstock/files/3984619/manuscript.pdf)

**Comment by agitter** [comment]
The simplification and elimination of redundancy in the build script is very nice.  Could you please test a CI or local DOCX build as well?

**Comment by dhimmel** [comment]
Here's the output from a local `BUILD_DOCX=true` build: [manuscript.docx](https://github.com/manubot/rootstock/files/3984807/manuscript.docx). Some formatting issues, but I don't think those are new.

<details>
<summary>Command output below</summary>

```
Exporting Word Docx manuscript
[INFO] Not rendering RawInline (Format "html") "<small>"
[INFO] Not rendering RawInline (Format "html") "<em>"
[INFO] Not rendering RawInline (Format "html") "</em>"
[INFO] Not rendering RawInline (Format "html") "</small>"
[INFO] Not rendering RawInline (Format "html") "<br>"
[INFO] Not rendering RawInline (Format "html") "<br>"
[INFO] Not rendering RawInline (Format "html") "<small>"
[INFO] Not rendering RawInline (Format "html") "</small>"
[INFO] Not rendering RawInline (Format "html") "<br>"
[INFO] Not rendering RawInline (Format "html") "<br>"
[INFO] Not rendering RawInline (Format "html") "<small>"
[INFO] Not rendering RawInline (Format "html") "</small>"
[INFO] Not rendering RawInline (Format "html") "<sup>"
[INFO] Not rendering RawInline (Format "html") "</sup>"
[INFO] Not rendering RawInline (Format "html") "<sub>"
[INFO] Not rendering RawInline (Format "html") "</sub>"
[INFO] Not rendering RawBlock (Format "html") "<!-- an arbitrary comment. visible in input, but not visible in output. -->"
[INFO] Fetching https://github.com/manubot/resources/raw/15493970f8882fce22bef829619d3fb37a613ba5/test/square.png...
[INFO] Fetching https://github.com/manubot/resources/raw/15493970f8882fce22bef829619d3fb37a613ba5/test/wide.png...
[INFO] Fetching https://github.com/manubot/resources/raw/15493970f8882fce22bef829619d3fb37a613ba5/test/tall.png...
[INFO] Fetching https://raw.githubusercontent.com/manubot/resources/master/test/vector.svg?sanitize=true...
[INFO] Not rendering RawInline (Format "html") "<!-- $colspan=\"2\" -->"
[INFO] Not rendering RawInline (Format "html") "<i class=\"fas fa-exclamation-triangle\">"
[INFO] Not rendering RawInline (Format "html") "</i>"
[INFO] Not rendering RawInline (Format "html") "<!-- $id=\"element_id\" class=\"some_class\" $style=\"color: #ad1457; margin-left: 40px;\" $disabled=\"true\" $title=\"a paragraph of text\" $data-color=\"red\" -->"
[INFO] Not rendering RawBlock (Format "html") "<!-- include the Font Awesome library, per: https://fontawesome.com/start -->"
[INFO] Not rendering RawInline (Format "html") "<link rel=\"stylesheet\" href=\"https://use.fontawesome.com/releases/v5.7.2/css/all.css\">"
[INFO] Not rendering RawInline (Format "html") "<i class=\"fas fa-check\">"
[INFO] Not rendering RawInline (Format "html") "</i>"
[INFO] Not rendering RawInline (Format "html") "<i class=\"fas fa-question\">"
[INFO] Not rendering RawInline (Format "html") "</i>"
[INFO] Not rendering RawInline (Format "html") "<i class=\"fas fa-star\">"
[INFO] Not rendering RawInline (Format "html") "</i>"
[INFO] Not rendering RawInline (Format "html") "<i class=\"fas fa-bell\">"
[INFO] Not rendering RawInline (Format "html") "</i>"
[INFO] Not rendering RawInline (Format "html") "<i class=\"fas fa-times-circle\">"
[INFO] Not rendering RawInline (Format "html") "</i>"
[INFO] Not rendering RawInline (Format "html") "<i class=\"fas fa-ellipsis-h\">"
[INFO] Not rendering RawInline (Format "html") "</i>"
[INFO] Not rendering RawInline (Format "html") "<i class=\"fas fa-scroll fa-lg\">"
[INFO] Not rendering RawInline (Format "html") "</i>"
[INFO] Not rendering RawInline (Format "html") "<br>"
[INFO] Not rendering RawInline (Format "html") "<i class=\"fas fa-info-circle fa-lg\">"
[INFO] Not rendering RawInline (Format "html") "</i>"
[INFO] Not rendering RawInline (Format "html") "<br>"
[INFO] Not rendering RawInline (Format "html") "<i class=\"fas fa-ban fa-lg\">"
[INFO] Not rendering RawInline (Format "html") "</i>"
[INFO] Not rendering RawInline (Format "html") "<br>"
[INFO] Not rendering RawBlock (Format "html") "<!-- Explicitly insert bibliography here -->"
```

</details>

**Comment by dhimmel** [comment]
One question is whether we should have additional default files for specific outputs.

For example, we could have `pandoc-defaults-html.yaml`, `pandoc-defaults-pdf-athena.yaml`, `pandoc-defaults-pdf-weasyprint.yaml`, and `pandoc-defaults-docx.yaml`. This would make it so the `pandoc` calls in `build.sh` only include `--defaults` (specified multiple times). This would make it easier to migrate the pandoc calls to our python package with the end goal of getting rid of `build.sh` and moving to a single command to process and build the manuscript.

**Comment by Nebucatnetzer** [comment]
Looks nice. Some mention in the README, USAGE might be good?
Another thing that came to mind, can one include or reference a file which overwrites the defaults?
In Ansible you can provide sane default  values which the user can easily overwrite and if he removes his customisation the default values get applied again. 
I haven't looked at the pandoc documentation yet.

Btw. please excuse that I'm a bit slow with responding.
I haven't been on my computer since Wednesday and doing code stuff on the phone or iPad is not that much fun :)

**Comment by dhimmel** [comment]
> can one include or reference a file which overwrites the defaults

I think there are two things a user may want to do:

1. discard an entire defaults file in favor of their own file. For example, do not use `html.yaml` at all and instead use `custom-html.yaml`

2. apply additional defaults files on top of the ones already specified by Manubot. This equates to specifying `--defaults` additional times on top of the Manubot builtin `--defaults` files. Pandoc has a behavior for merging defaults, but this is not entirely consistent and well documented yet as per https://github.com/jgm/pandoc/issues/5988 (will likely improve).

I'd like to look into whether we can accomplish 2 without requiring any changes to `build.sh`. For example, whether we could do `--defaults=html*.yaml`

> please excuse that I'm a bit slow with responding

@Nebucatnetzer you're under no obligation to respond at all... so nothing to excuse :smiley_cat:

**Comment by Nebucatnetzer** [comment]
> I think there are two things a user may want to do:
> 
>     1. discard an entire defaults file in favor of their own file. For example, do not use `html.yaml` at all and instead use `custom-html.yaml`
> 
>     2. apply additional defaults files on top of the ones already specified by Manubot. This equates to specifying `--defaults` additional times on top of the Manubot builtin `--defaults` files. Pandoc has a behavior for merging defaults, but this is not entirely consistent and well documented yet as per [jgm/pandoc#5988](https://github.com/jgm/pandoc/issues/5988) (will likely improve).
> 
> 
> I'd like to look into whether we can accomplish 2 without requiring any changes to `build.sh`. For example, whether we could do something equivalent to `--defaults=html*.yaml` (or do this in our Python tooling eventually)
> 
> > please excuse that I'm a bit slow with responding
> 
> @Nebucatnetzer you're under no obligation to respond at all... so nothing to excuse smiley_cat

Aye something like `--defaults=pandoc-custom/*.yaml` might work, I reckon a sub directory would be a bit cleaner.
Maybe we would've to include an empty YAML file so that it doesn't trip when there aren't any customisations present.

As for 1. maybe we could make the build.sh a bit more complex so that if it finds a YAML with the exact same name as a defaults file (e.g. html.yaml) it would replace the defaults and if it has a different name (e.g. custom-html.yml) it would be added on top?

**Comment by dhimmel** [comment]
> As for 1. maybe we could make the build.sh a bit more complex

I looked briefly into using bash / shell commands to detect files matching a collection of patterns and then transforming each file into a pandoc `--defaults` argument. There didn't seem to be an easy / readable solution.

This makes me lean towards implementing this logic in Python. So the proposal would be:

1. merge this PR with the current design where all `--defaults` files must be manually specified when calling pandoc.
2. work on python code to detect default files (either additional or replacement files) in the `manubot` [package](https://github.com/manubot/manubot).

**Comment by Nebucatnetzer** [comment]
Sounds good to me :)

File: build/build.sh [hunk 1]
```diff
@@ -19,42 +19,15 @@ manubot process \
...
-  --include-after-body=build/plugins/lightbox.html \
-  --include-after-body=build/plugins/attributes.html \
-  --mathjax \
-  --variable math="" \
-  --include-after-body=build/plugins/math.html \
-  --include-after-body=build/plugins/hypothesis.html \
-  --include-after-body=build/plugins/analytics.html \
-  --output=output/manuscript.html \
-  "$INPUT_PATH"
+  --defaults=build/pandoc-defaults/common.yaml \
```
Comment (agitter) [inline build/build.sh]: The proposal in https://github.com/manubot/rootstock/pull/261#issuecomment-568220679 would give more flexibility for specifying where the defaults files reside.

Comment (dhimmel) [reply build/build.sh]: @agitter are you suggesting that we allow setting the directory like:

```
PANDOC_DEFAULTS_DIR="${PANDOC_DEFAULTS_DIR:-build/pandoc-defaults}"
``

This means there's a single environment variable and that users could do:

```shell
cp build/pandoc-defaults build/custom-pandoc-defaults
# edit files in build/custom-pandoc-defaults
export PANDOC_DEFAULTS_DIR=build/custom-pandoc-defaults
```

Or do you think each file should have it's own variable like `PANDOC_DEFAULTS_COMMON_PATH`?

Comment (agitter) [reply build/build.sh]: I propose that initially we do:
```
PANDOC_DEFAULTS_DIR="${PANDOC_DEFAULTS_DIR:-build/pandoc-defaults}"
```
Your example customization above looks pretty straightforward.

That may end up being a placeholder depending what changes as part of
> work on python code to detect default files (either additional or replacement files) in the manubot package.

**Pull request merged**